File: PyProject/leetcode/history/merge_sort.py
from typing import List


def merge_sort(arr: List[int]):
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]
        R = arr[mid:]
        merge_sort(L)
        merge_sort(R)

        arr.clear()

        while len(L) > 0 and len(R) > 0:
            if L[0] < R[0]:
                arr.append(L[0])
                L.pop(0)
            else:
                arr.append(R[0])
                R.pop(0)
        for i in L:
            arr.append(i)
        for j in R:
            arr.append(j)


if __name__ == '__main__':
    a = [1, 4, 3, 2, 8, 3, 6]
    merge_sort(a)
    print(a)

File: GoProject/leetcode/sort/20210321/mergeSort.go
package main

import "fmt"

func mergeSort(nums []int) []int {
	if len(nums) <= 1 {
		return nums
	}
	l, r := 0, 0

	mid := len(nums) / 2
	left := mergeSort(nums[0:mid])
	right := mergeSort(nums[mid:len(nums)])

	ll, lr := len(left), len(right)

	result := make([]int, 0)

	for l < ll && r < lr {
		if left[l] < right[r] {
			result = append(result, left[l])
			l++
		} else {
			result = append(result, right[r])
			r++
		}
	}

	result = append(result, left[l:]...)
	result = append(result, right[r:]...)

	return result

}

func main() {
	arr := []int{10, 7, 8, 9, 1, 5}
	result := mergeSort(arr)
	fmt.Println(result)
}
